vault/api_redirect.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package vault
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	url2 "net/url"
+	"strings"
+	"sync"
+
+	"github.com/armon/go-radix"
+)
+
+type apiRedirectRegistry struct {
+	lock  sync.Mutex
+	paths *radix.Tree
+}
+
+func NewAPIRedirects() *apiRedirectRegistry {
+	return &apiRedirectRegistry{
+		paths: radix.New(),
+	}
+}
+
+func (reg *apiRedirectRegistry) TryRegister(ctx context.Context, core *Core, mountUUID, src, dest string) error {
+	if strings.HasPrefix(dest, "/") {
+		return errors.New("redirect targets must be relative")
+	}
+	reg.lock.Lock()
+	defer reg.lock.Unlock()
+	_, _, found := reg.paths.LongestPrefix(src)
+	if found {
+		return fmt.Errorf("api redirect conflict for %s", src)
+	}
+	reg.paths.Insert(src, &APIRedirect{
+		c:         core,
+		mountUUID: mountUUID,
+		prefix:    dest,
+	})
+	return nil
+}
+
+func (reg *apiRedirectRegistry) Find(path string) (*APIRedirect, string) {
+	s, a, found := reg.paths.LongestPrefix(path)
+	if found {
+		remaining := strings.TrimPrefix(path, s)
+		if len(remaining) > 0 {
+			switch remaining[0] {
+			case '/':
+				remaining = remaining[1:]
+			case '?':
+			default:
+				// This isn't an exact path match
+				return nil, ""
+			}
+		}
+		return a.(*APIRedirect), remaining
+	}
+	return nil, ""
+}
+
+func (reg *apiRedirectRegistry) DeregisterMount(uuid string) {
+	reg.lock.Lock()
+	defer reg.lock.Unlock()
+	reg.paths.Walk(func(k string, v interface{}) bool {
+		r := v.(*APIRedirect)
+		if r.mountUUID == uuid {
+			reg.paths.Delete(k)
+			return true
+		}
+		return false
+	})
+}
+
+func (reg *apiRedirectRegistry) DeregisterSource(uuid, src string) bool {
+	reg.lock.Lock()
+	defer reg.lock.Unlock()
+	var found bool
+	reg.paths.Walk(func(k string, v interface{}) bool {
+		r := v.(*APIRedirect)
+		if r.mountUUID == uuid && k == src {
+			found = true
+			reg.paths.Delete(k)
+			return true
+		}
+		return false
+	})
+	return found
+}
+
+type APIRedirect struct {
+	c             *Core
+	mountUUID     string
+	prefix        string
+	isPrefixMatch bool
+}
+
+func (a *APIRedirect) Destination(remaining string) (string, error) {
+	var destPath string
+	if a.c == nil {
+		// Just for testing
+		destPath = a.prefix
+	} else {
+		m := a.c.mounts.findByMountUUID(a.mountUUID)
```

Comment:
Could we not use the existing `a.c.router.MatchingMountByUUID` method instead of defining a new way?

Reply:
Oh, didn't know that existed.  I guess that's fine if the cache is up to date.